### PR TITLE
Game of Amazons API troubles

### PR DESCRIPTION
**Pull request by selfsim**
Hello community,

I am a C++/open_spiel newb and am hoping to get some pointers on how to correctly set up my implementation of "The Game of Amazons" to adhere to open_spiel's Game API so that I can train some reinforcement learning agents.

So far I have followed the [adding a game docs](https://openspiel.readthedocs.io/en/latest/developer_guide.html#adding-a-game) and am able to load my game. However, my implementation fails other open_spiel provided tests. Following is my debugging attempt:

I compiled in debug mode to use gdb to hunt the segfault:

I ran "gdb amazons_test" from the build/games directory.

Following is the trace:
 
image

The issue seems to be my misunderstanding of how open_spiel encodes actions. I understood Action as a 64 bit binary string which encodes the actions players can take in the game. I wrote an encoder/decoder in my amazons.cc file. This seems wrong (see image), it looks like Game expects the actions to be index-like. I'm hoping someone can point me in the right way on how to fix that.

Here is [a link to the code](https://github.com/deepmind/open_spiel/blob/c7da3c74d09d656f4ac091801bbd48782c1fda2c/open_spiel/games/amazons.cc#L373) where I define LegalActions within amazons.cc.

Additionally, I am interested in learning how to make/compile a single test file so that I can speed up my debugging.

Sorry for the newb questions.

Any help is appreciated :-).

-selfsim

### CODE REVIEW

**Comment by selfsim** [comment]
> Thanks for your pull request. It looks like this may be your first contribution to a Google open source project (if not, look below for help). Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
>  **Please visit https://cla.developers.google.com/ to sign.**
> 
> Once you've signed (or fixed any issues), please reply here with `@googlebot I signed it!` and we'll verify it.
> 
> #### What to do if you already signed the CLA
> ##### Individual signers
> * It's possible we don't have your GitHub username or you're using a different email address on your commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> 
> ##### Corporate signers
> * Your company has a Point of Contact who decides which employees are authorized to participate. Ask your POC to be added to the group of authorized contributors. If you don't know who your Point of Contact is, direct the Google project maintainer to [go/cla#troubleshoot](http://go/cla#troubleshoot) ([Public version](https://opensource.google/docs/cla/#troubleshoot)).
> * The email used to register you as an authorized contributor must be the email used for the Git commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * The email used to register you as an authorized contributor must also be [attached to your GitHub account](https://github.com/settings/emails).
> 
>  **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fdeepmind%2Fopen_spiel%2Fpull%2F537) for more info**.

@googlebot I signed it!

**Comment by michalsustr** [comment]
Hi, not sure what the problem is exactly from a quick glance, but some suggestions:

1) The legal action mask caused segfault, because it's trying to mask actions outside of its mask size. Check that the number of distinct actions / chance outcomes upper bounds your maximum action values, so that they can be used for creating large enough arrays. 

2) Encoding the actions as three 8-bit numbers will yield 24bits, which are quite large values. I am not sure how the action mask got invoked, because I don't think it's part of some standard tests (but I might be wrong). Anyway, you might want to encode the actions in more compressed format so they don't occupy as much space. Check out how this is done in Chess as multiplication/division with board sizes for encoding/decoding moves.

**Comment by lanctot** [comment]
Hi @selfsim, yes the actions need to be within a specfic range: 0 <= action < game->NumDistinctActions(). They can't be arbitrary 64-bit integers, because many algorithms need to know the right-hand side...  e.g. if you're doing neural networks you need to have a finite width policy head. 

Probably the best is to create the encoder / decoder so that everything fits with in the range. If that's too hard, then you can maintain an internal map back and forth, building it only at game construction time (see blotto as an example). For Amazons I think it'd be easy enough to do the former, which would be preferred. We can help you with that, as there are many places in the code that do this. But also, if that leads to too many actions then we can split the subturns into two separate turns (i.e. move first, then shoot)-- but this is probably not necessary (a few hundred or a thousand distinct actions is ok).

**Comment by selfsim** [comment]
> Hi, not sure what the problem is exactly from a quick glance, but some suggestions:
> 
>     1. The legal action mask caused segfault, because it's trying to mask actions outside of its mask size. Check that the number of distinct actions / chance outcomes upper bounds your maximum action values, so that they can be used for creating large enough arrays.
> 
>     2. Encoding the actions as three 8-bit numbers will yield 24bits, which are quite large values. I am not sure how the action mask got invoked, because I don't think it's part of some standard tests (but I might be wrong). Anyway, you might want to encode the actions in more compressed format so they don't occupy as much space. Check out how this is done in Chess as multiplication/division with board sizes for encoding/decoding moves.

Hi, @michalsustr  The action mask was invoked through the RandomSimTest found  [here](https://github.com/deepmind/open_spiel/blob/c7da3c74d09d656f4ac091801bbd48782c1fda2c/open_spiel/games/amazons_test.cc#L26). 

Thanks for your suggestion to compress the actions, I will take a look at that.

**Comment by selfsim** [comment]
> Hi @selfsim , I refreshed my memory of Amazons and I have some recommendations. I forgot it was a 10x10 board, too.
> 
> The lowest I can get the actions down is 4_8_9_8_9 = 20736, if you consider an action both for movement and the arrow. I get this from: 4 choices of the amazon to move, 8 directions, 9 squares in the direction, 8 directions (arrow shot), 9 squares in the direction (arrow shot). That's a lot, it would be the largest action game we have and it's almost surely an issue if you want to do anything with vanilla policy nets or q-networks.
> 
> I'd recommend the subturns: each player chooses a movement (4_8_9 = 288 actions) and then an arrow shot (72 actions). This is much more manageable if you're planning to do any RL.
> 
> I don't like that the 4 includes having to label each amazon or make assumptions about how you choose them to act (based on the state) but if you don't do that, then it gets big fast (100 * 8 * 9 = 7200 actions for movement alone).
> 
> Of course, this is just the representation of the game. You could use smarter things on the networks side like having separate output heads that encode a full action and learn those. It might depend what kind of research you are looking to do with the game.
> 
> If there are no other things affecting the design choices, I'd say the 288+72 route is a decent way to go. The 20736 actions could work too with the understanding that doing basic RL would be likely infeasible unless you do search / value function learning or some kind of specific network architecture, but 20736 actions is quite a big range.
> 
> Thoughts?

Hi @lanctot , thanks for your reply. The problem of the large action space is something that I thought about too, and it led me to hierarchical RL, which I am learning more about. I definitely like the idea of splitting the action into move/shoot, and will put it in my list of future improvements. For now, I'll just try to keep things as simple as possible and improve the encoding/decoding to fit the range of (0, game->NumDistinctActions()). 

The purpose of this project is to learn as much as I can about RL while building a bot which plays Amazons relatively well. My plan is to start at a low resolution, get some results, and then incrementally improve upon them, by making changes to the game implementation and playing around with different RL algorithms/approaches, with an end goal of gaining some intuition.

**Comment by lanctot** [comment]
Perfect! There are general functions for ranking and unranking integers in mixed bases to encode actions. I believe Breakthrough uses these helpers. 

If you would try something like DQN or AlphaZero then I'd say you will really want to try to keep the NumDistinctActions() to be as low as possible, otherwise you will be stuck (i.e. 20k actions will be a big bottleneck).

Feel free to use this PR to iterate, and ask us questions along the way!

**Comment by selfsim** [comment]
Hi @lanctot, I've decided to implement sub-turns directly and found success - the implementation is passing RandomSimTests.
Going off of your idea, I broke the turn into amazon select, amazon move, and arrow shot, bounding the action space to rows * columns per move. I think its a good start, I will refactor and optimize the implementation as my C++ knowledge grows.

Are there any other tests I should try my implementation against before moving forward?

Otherwise I'm going to start tackling the C++ RL examples.

Cheers!

**Comment by lanctot** [comment]
Wow, that's amazing progress! Those changes were really non-trivial, very nice. 

It's always helpful to have a few specific tests that walk through the game and test some corner cases that might arise. They don't have to be extensive, but it'd be good if there's some confidence that it's fully correct. You can also hand-construct the states somehow and test the cases independently. 

You can look into the backgammon_test to see a few of these. You don't need many of them to get started, maybe just 2-3 of the most probable cases not handled. (Backgammon is notorious for these tricky corner cases, I imagine there's far fewer in Amazons). Like maybe you want to just walk a specific trajectory and check that you're not able to move through arrows or shoot through other arrows / pieces. I'm not super familiar with the game so you can decide on the specific tests. 

Other than that someone will do a pass through to review the code and ask you to make some changes -- once that's done, we'll pull it in.

**Comment by lanctot** [comment]
> Otherwise I'm going to start tackling the C++ RL examples.

Not sure what you had in mind, but we actually don't have many C++ RL examples. Mostly just tabular ones (Q-learning and Sarsa). We do have two AlphaZero's, one based on TF and one based on PyTorch. Both require installing optional dependencies, and the TF one can be especially tricky. 

It'd be really great to have an entirely C++ DQN based on one of these ML libraries as well. In fact, I think it'd mostly be straight-forward to translate our PyTorch DQN to C++. That is a project I'd really truly love you to take on if you're interested, that would be such a unique and fun addition to the library. I also have someone on our side who might be interested in collaborating with you on that, if you're so inclined!

I'm happy to help suggest where to go from here if you tell me a bit more about what you have in mind.

**Comment by selfsim** [comment]
Hi @lanctot,

great, I will think of some ways to verify that the implementation is good.

I'm interested in using the TF AlphaZero implementation to get a bot up and running. Could you comment a little about how I might move forward in this? Is building with the TF flag set insufficient?

As for the C++ DQN implementation.. that sounds interesting. I'd happily work on something like that. I think it would be a great learning experience.

Cheers!

**Comment by lanctot** [comment]
> I'm interested in using the TF AlphaZero implementation to get a bot up and running. Could you comment a little about how I might move forward in this? Is building with the TF flag set insufficient?

Best thing is follow the instructions in the README here: https://github.com/deepmind/open_spiel/tree/master/open_spiel/algorithms/alpha_zero. TensorflowCC will get CMake-compatible build TF builds, and you need to use the @mrdaliri 's fork linked in the README above. 

The main problem is that TF is not easy to compile from source, and there's been a number of people who have run into problems (for example we got a recent issue today, see here: https://github.com/deepmind/open_spiel/issues/539 and a few weeks back that we never solved: https://github.com/deepmind/open_spiel/issues/482). It's just impossible for us to support this other than through suggestions, because we just don't have the time. Whereas I know a few people already using the Libtorch-based C++ AlphaZero (here: https://github.com/deepmind/open_spiel/tree/master/open_spiel/algorithms/alpha_zero_torch) and it just works because Libtorch is so much easier to install.

**Comment by selfsim** [comment]
@lanctot Thanks for the heads up. I will take a crack at the Libtorch C++ AlphaZero this week.

As for the DQN implementation, could you open an issue? Then the person who you were mentioning that was interested in collaborating and I could get into touch.

**Comment by lanctot** [comment]
> As for the DQN implementation, could you open an issue? Then the person who you were mentioning that was interested in collaborating and I could get into touch.

Sure, here it is: https://github.com/deepmind/open_spiel/issues/542. The person I was thinking about is Satyaki Upadhyay (@satyaki3794).

**Comment by lanctot** [comment]
@selfsim can you add the Amazon tests e discussed? It would be great to review then import this PR so we can be sure it is correct, before you investigate significant time using it.

**Comment by selfsim** [comment]
@lanctot will do. Will ping when they are pushed up

**Comment by selfsim** [comment]
Hi @lanctot, I've come up with a few simple tests for LegalActions(), the workhorse of my implementation. Given the successful reward reporting of random simulation outcomes and these, I think the implementation is solid. 

Re: your suggestion for a walkthrough checking whether or not an amazon can move through a block or another amazon: my LegalActions() generation seems good, so no player would be able to select an illegal action like these.

Let me know, I can write more if you have doubts.

Cheers!

**Comment by lanctot** [comment]
This should be fine! Thanks.  I am out-of-office currently, but can take a deeper look next week. 

@tewalds @elkhrt @michalsustr do one you want to take a look at this in the mean time?

**Comment by lanctot** [comment]
@selfsim can you add Amazons to the games list `docs/games.md`? We should be importing this soon.

**Comment by bartdevylder** [comment]
Hi @selfsim, thanks for your contribution! The code seems almost ready to import, could you just add an integration test as described here: https://openspiel.readthedocs.io/en/latest/developer_guide.html#adding-a-game -> "Update Python integration tests"? I could generate this myself but it's good to check whether the generated game playthrough matches what you'd expect given the rules of the game and your implementation.

**Comment by lanctot** [comment]
@selfsim We are planning a release of OpenSpiel, possibly the big one (v1.0). It would be really cool to announce Amazons in the release. Do you think you'll have time to look into this before the release (end of May or early June)?

**Comment by selfsim** [comment]
@lanctot @bartdevylder Hi guys, I'm just on break from uni currently and taking a few days off after the final storm. I will update the `games.md` and write an integration test soon.

**Comment by lanctot** [comment]
Hi @selfsim, I am now thinking of an August release and just wondering whether you still have time and intend to finish up the last remaining things for this Amazons game implementation. Would be great to include it with the release.

**Comment by lanctot** [comment]
Hi @selfsim, closing this for now since we have not heard from you. Feel free to reopen to finish the last steps if/when you have time, and we can still import it when ready.

**Comment by lanctot** [comment]
Welcome back :)

Ok yes I see the addition of amazons to pyspiel_test.py. @bartdevylder, was there something else you wanted in addition? I believe this satisfies that part of adding a game that you pointed out.

**Comment by lanctot** [comment]
There seems to be a problem with the cla now, though.. maybe it's outdated? Can you look into it.. we can't import until that is fixed. You might have to submit a new PR.

**Comment by lanctot** [comment]
But also some tests are now failing. Best thing to do at this point is to copy these files and submit it as a new PR, and we can go from there.

**Comment by bartdevylder** [comment]
Hi selfsim, welcome back! 
And thanks for the changes. It seems only the last step (9) in https://openspiel.readthedocs.io/en/latest/developer_guide.html#adding-a-game still needs to be executed: it will generate a 'playthrough' file in integration_tests/playthroughs
(admittedly the description "Update Python integration tests" is a bit vague, it will be updated)
